### PR TITLE
Add payload support to `onMessage` method

### PR DESCRIPTION
**Pull request by dilanSachi**
## Purpose
$subject
Related to https://github.com/ballerina-platform/ballerina-standard-library/issues/2880

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [ ] Updated the spec

### CODE REVIEW

File: ballerina/tests/data_binding_publish_tests.bal [hunk 1]
```diff
@@ -617,6 +636,637 @@ service object {
...
+    }
+}
 
+Service consumerServiceBytesPayloadPublish =
+@ServiceConfig {
+    subject: DATA_BINDING_PUBLISH_SUBJECT
+}
+service object {
+    remote function onMessage(byte[] payload) {
+        byte[] messageContent = <@untainted> payload;
```
Comment (Bhashinee) [inline ballerina/tests/data_binding_publish_tests.bal:733]:
```suggestion
        byte[] messageContent = payload;
```

File: ballerina/tests/data_binding_publish_tests.bal [hunk 2]
```diff
@@ -31,9 +31,28 @@ boolean receivedBooleanValuePublish = false;
 Person? receivedPersonValuePublish = ();
 map<Person>? receivedMapValuePublish = ();
 table<Person>? receivedTableValuePublish = ();
+string receivedStringPayloadValuePublish = "";
+string receivedBytesPayloadValuePublish = "";
```
Comment (Bhashinee) [inline ballerina/tests/data_binding_publish_tests.bal:35]: Do we need these 2 variables? Didn't see any usage.

File: ballerina/tests/data_binding_publish_tests.bal [hunk 3]
```diff
@@ -617,6 +636,637 @@ service object {
...
+        if sub is Listener {
+            checkpanic sub.attach(consumerServiceStringPayloadPublish);
+            checkpanic sub.'start();
+            checkpanic newClient->publishMessage({ content: messageToSend, subject: DATA_BINDING_PUBLISH_SUBJECT });
+            int timeoutInSeconds = 120;
+            // Test fails in 2 minutes if it is failed to receive the message
+            while timeoutInSeconds > 0 {
+                if receivedStringValuePublish !is "" {
+                    string receivedMessage = receivedStringValuePublish;
+                    test:assertEquals(receivedMessage, messageToSend, msg = "Message received does not match.");
```
Comment (Bhashinee) [inline ballerina/tests/data_binding_publish_tests.bal:660]:
```suggestion
                    test:assertEquals(receivedStringValuePublish, messageToSend, msg = "Message received does not match.");
```

File: ballerina/tests/data_binding_publish_tests.bal [hunk 4]
```diff
@@ -617,6 +636,637 @@ service object {
...
+        if sub is Listener {
+            checkpanic sub.attach(consumerServiceBytesPayloadPublish);
+            checkpanic sub.'start();
+            checkpanic newClient->publishMessage({ content: messageToSend.toBytes(), subject: DATA_BINDING_PUBLISH_SUBJECT });
+            int timeoutInSeconds = 120;
+            // Test fails in 2 minutes if it is failed to receive the message
+            while timeoutInSeconds > 0 {
+                if receivedBytesValuePublish !is "" {
+                    string receivedMessage = receivedBytesValuePublish;
+                    test:assertEquals(receivedMessage, messageToSend, msg = "Message received does not match.");
```
Comment (Bhashinee) [inline ballerina/tests/data_binding_publish_tests.bal:708]:
```suggestion
                    test:assertEquals(receivedBytesValuePublish, messageToSend, msg = "Message received does not match.");
```

File: ballerina/tests/data_binding_publish_tests.bal [hunk 5]
```diff
@@ -617,6 +636,637 @@ service object {
...
+        if sub is Listener {
+            checkpanic sub.attach(consumerServiceXmlPayloadPublish);
+            checkpanic sub.'start();
+            checkpanic newClient->publishMessage({ content: messageToSend, subject: DATA_BINDING_PUBLISH_SUBJECT });
+            int timeoutInSeconds = 120;
+            // Test fails in 2 minutes if it is failed to receive the message
+            while timeoutInSeconds > 0 {
+                if receivedXmlPayloadValuePublish !is "" {
+                    string receivedMessage = receivedXmlPayloadValuePublish;
+                    test:assertEquals(receivedMessage, messageToSend.toString(), msg = "Message received does not match.");
```
Comment (Bhashinee) [inline ballerina/tests/data_binding_publish_tests.bal:760]:
```suggestion
                    test:assertEquals(receivedXmlPayloadValuePublish, messageToSend.toString(), msg = "Message received does not match.");
```

File: ballerina/tests/data_binding_publish_tests.bal [hunk 6]
```diff
@@ -617,6 +636,637 @@ service object {
...
+        if sub is Listener {
+            checkpanic sub.attach(consumerServiceJsonPayloadPublish);
+            checkpanic sub.'start();
+            checkpanic newClient->publishMessage({ content: messageToSend, subject: DATA_BINDING_PUBLISH_SUBJECT });
+            int timeoutInSeconds = 120;
+            // Test fails in 2 minutes if it is failed to receive the message
+            while timeoutInSeconds > 0 {
+                if receivedJsonPayloadValuePublish !is () {
+                    json receivedMessage = receivedJsonPayloadValuePublish;
+                    test:assertEquals(receivedMessage, messageToSend, msg = "Message received does not match.");
```
Comment (Bhashinee) [inline ballerina/tests/data_binding_publish_tests.bal:811]:
```suggestion
                    test:assertEquals(receivedJsonPayloadValuePublish, messageToSend, msg = "Message received does not match.");
```

File: ballerina/tests/data_binding_publish_tests.bal [hunk 7]
```diff
@@ -617,6 +636,637 @@ service object {
...
+        if sub is Listener {
+            checkpanic sub.attach(consumerServiceIntPayloadPublish);
+            checkpanic sub.'start();
+            checkpanic newClient->publishMessage({ content: messageToSend, subject: DATA_BINDING_PUBLISH_SUBJECT });
+            int timeoutInSeconds = 120;
+            // Test fails in 2 minutes if it is failed to receive the message
+            while timeoutInSeconds > 0 {
+                if receivedIntPayloadValuePublish !is 0 {
+                    int receivedMessage = receivedIntPayloadValuePublish;
+                    test:assertEquals(receivedMessage, messageToSend, msg = "Message received does not match.");
```
Comment (Bhashinee) [inline ballerina/tests/data_binding_publish_tests.bal:859]:
```suggestion
                    test:assertEquals(receivedIntPayloadValuePublish, messageToSend, msg = "Message received does not match.");
```

Comment (Bhashinee) [reply ballerina/tests/data_binding_publish_tests.bal:859]: Please check other places as well.

File: ballerina/tests/data_binding_publish_tests.bal [hunk 8]
```diff
@@ -617,6 +636,637 @@ service object {
...
+}
+public function testDataBindingStringPayloadPublish() {
+    string messageToSend = "Testing Consumer Service With String Data Binding.";
+    Client? newClient = clientObj;
+    if newClient is Client {
+        Listener? sub = listenerObj;
+        if sub is Listener {
+            checkpanic sub.attach(consumerServiceStringPayloadPublish);
+            checkpanic sub.'start();
+            checkpanic newClient->publishMessage({ content: messageToSend, subject: DATA_BINDING_PUBLISH_SUBJECT });
```
Comment (Bhashinee) [inline ballerina/tests/data_binding_publish_tests.bal:654]: Without using `checkpanic`, let's use `check` and return an error from the test function.

Comment (Bhashinee) [reply ballerina/tests/data_binding_publish_tests.bal:654]: Let's fix this in other places as well.

File: native/src/main/java/io/ballerina/stdlib/nats/Utils.java [hunk 9]
```diff
@@ -126,17 +128,35 @@ public static Object getValueWithIntendedType(Type type, byte[] value) throws BE
                     return ValueCreator.createArrayValue(value);
                 case RECORD_TYPE_TAG:
                     return CloneWithType.convert(type, JsonUtils.parse(strValue));
+                case UNION_TAG:
+                    if (hasStringType((UnionType) type)) {
+                        return StringUtils.fromString(strValue);
+                    }
+                    return getValueFromJson(type, strValue);
```
Comment (Bhashinee) [inline native/src/main/java/io/ballerina/stdlib/nats/Utils.java:135]: Don't we need to fall through? To address scenarios union types like `byte[]|string payload`? Not sure whether this is a valid use case in messaging :)

Comment (dilanSachi) [reply native/src/main/java/io/ballerina/stdlib/nats/Utils.java:135]: Its covered using `getValueFromJson(type, strValue);`

File: ballerina/tests/data_binding_publish_tests.bal (hunk 8 above)
Comment (dilanSachi) [reply ballerina/tests/data_binding_publish_tests.bal]: Fixed in https://github.com/ballerina-platform/module-ballerinax-nats/pull/427/commits/433577ffabea99c86a5ec34828b469bce521fc85

File: ballerina/tests/data_binding_publish_tests.bal (hunk 7 above)
Comment (dilanSachi) [reply ballerina/tests/data_binding_publish_tests.bal]: Fixed in https://github.com/ballerina-platform/module-ballerinax-nats/pull/427/commits/433577ffabea99c86a5ec34828b469bce521fc85

File: ballerina/tests/data_binding_publish_tests.bal (hunk 6 above)
Comment (dilanSachi) [reply ballerina/tests/data_binding_publish_tests.bal]: Fixed in https://github.com/ballerina-platform/module-ballerinax-nats/pull/427/commits/433577ffabea99c86a5ec34828b469bce521fc85

File: ballerina/tests/data_binding_publish_tests.bal (hunk 5 above)
Comment (dilanSachi) [reply ballerina/tests/data_binding_publish_tests.bal]: Fixed in https://github.com/ballerina-platform/module-ballerinax-nats/pull/427/commits/433577ffabea99c86a5ec34828b469bce521fc85

File: ballerina/tests/data_binding_publish_tests.bal (hunk 4 above)
Comment (dilanSachi) [reply ballerina/tests/data_binding_publish_tests.bal]: Fixed in https://github.com/ballerina-platform/module-ballerinax-nats/pull/427/commits/433577ffabea99c86a5ec34828b469bce521fc85

File: ballerina/tests/data_binding_publish_tests.bal (hunk 3 above)
Comment (dilanSachi) [reply ballerina/tests/data_binding_publish_tests.bal]: Fixed in https://github.com/ballerina-platform/module-ballerinax-nats/pull/427/commits/433577ffabea99c86a5ec34828b469bce521fc85

File: ballerina/tests/data_binding_publish_tests.bal (hunk 2 above)
Comment (dilanSachi) [reply ballerina/tests/data_binding_publish_tests.bal:35]: Missed to add these to the test cases
Fixed in https://github.com/ballerina-platform/module-ballerinax-nats/pull/427/commits/433577ffabea99c86a5ec34828b469bce521fc85

File: ballerina/tests/data_binding_publish_tests.bal (hunk 1 above)
Comment (dilanSachi) [reply ballerina/tests/data_binding_publish_tests.bal]: Fixed in https://github.com/ballerina-platform/module-ballerinax-nats/pull/427/commits/433577ffabea99c86a5ec34828b469bce521fc85

**Review by Bhashinee — APPROVED** [review]

**Pull request merged**